File: scripts/query-token-rpc.mjs
#!/usr/bin/env node
/**
 * Solana RPC token query script.
 * Usage:  node scripts/query-token-rpc.mjs [mint-address]
 *
 * Queries a Solana devnet token for:
 *   - Total supply
 *   - Largest holders
 *   - Recent transaction signatures
 *   - Token account (holder) count
 *
 * Default: USDC devnet (Gh9ZwEmdLJ8DscKNTkTqPbNwLNNBjuSzaG9Vp2KGtKJr)
 */

const RPC = "https://api.devnet.solana.com";
const DEFAULT_MINT = "Gh9ZwEmdLJ8DscKNTkTqPbNwLNNBjuSzaG9Vp2KGtKJr";

let id = 1;

async function rpc(method, params = [], retries = 3) {
  for (let i = 0; i < retries; i++) {
    try {
      const res = await fetch(RPC, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ jsonrpc: "2.0", id: id++, method, params }),
      });
      const json = await res.json();
      if (json.error) {
        if (json.error.message?.includes("Too many requests") && i < retries - 1) {
          await new Promise(r => setTimeout(r, 1000 * (i + 1)));
          continue;
        }
        throw new Error(json.error.message);
      }
      return json.result;
    } catch (e) {
      if (i < retries - 1) {
        await new Promise(r => setTimeout(r, 1000 * (i + 1)));
        continue;
      }
      throw e;
    }
  }
}

async function main() {
  const mint = process.argv[2] || DEFAULT_MINT;
  console.log(`\n=== Solana Devnet RPC Token Query ===`);
  console.log(`Mint: ${mint}\n`);

  // 1. Supply
  console.log("1. Supply...");
  const info = await rpc("getAccountInfo", [mint, { encoding: "jsonParsed" }]);
  const p = info?.value?.data?.parsed?.info;
  if (p) {
    const supply = Number(p.supply) / 10 ** p.decimals;
    console.log(`   Decimals:   ${p.decimals}`);
    console.log(`   Supply:     ${supply.toLocaleString(undefined, { maximumFractionDigits: 2 })}`);
  }

  // 2. Largest holders
  console.log("\n2. Largest holders...");
  await new Promise(r => setTimeout(r, 500));
  try {
    const largest = await rpc("getTokenLargestAccounts", [mint]);
    (largest.value || []).slice(0, 5).forEach((a, i) => {
      console.log(`   [${i + 1}] ${a.address.slice(0, 12)}...  ${Number(a.uiAmount).toLocaleString(undefined, { maximumFractionDigits: 2 })}`);
    });
  } catch (e) {
    console.log(`   Skipped: ${e.message}`);
  }

  // 3. Recent signatures
  console.log("\n3. Recent transaction signatures...");
  await new Promise(r => setTimeout(r, 500));
  try {
    const sigs = await rpc("getSignaturesForAddress", [mint, { limit: 5 }]);
    console.log(`   Found: ${sigs.length} sigs`);
    sigs.slice(0, 3).forEach(s => console.log(`   ${s.signature.slice(0, 20)}...  slot=${s.slot}`));
  } catch (e) {
    console.log(`   ${e.message}`);
  }

  console.log("\n=== Done ===");
}

main().catch(e => { console.error("Fatal:", e.message); process.exit(1); });

File: scripts/query-token-rpc.py
#!/usr/bin/env python3
"""
Solana RPC token query script.
Usage:  python scripts/query-token-rpc.py [mint-address]

Queries a Solana devnet token for:
  - Total supply
  - Largest holders
  - Recent transaction signatures
  - Token account (holder) count

Default: USDC devnet (Gh9ZwEmdLJ8DscKNTkTqPbNwLNNBjuSzaG9Vp2KGtKJr)
"""

import json
import sys
import time
import urllib.request

RPC = "https://api.devnet.solana.com"
DEFAULT_MINT = "Gh9ZwEmdLJ8DscKNTkTqPbNwLNNBjuSzaG9Vp2KGtKJr"


def rpc(method, params=None, retries=3):
    params = params or []
    for attempt in range(retries):
        try:
            payload = json.dumps({"jsonrpc": "2.0", "id": 1, "method": method, "params": params}).encode()
            req = urllib.request.Request(RPC, data=payload, headers={"Content-Type": "application/json"})
            with urllib.request.urlopen(req, timeout=15) as resp:
                result = json.loads(resp.read().decode())
            if "error" in result:
                msg = result["error"].get("message", "")
                if "Too many requests" in msg and attempt < retries - 1:
                    time.sleep(1 * (attempt + 1))
                    continue
                raise RuntimeError(msg)
            return result["result"]
        except Exception as e:
            if attempt < retries - 1:
                time.sleep(1 * (attempt + 1))
                continue
            raise


def main():
    mint = sys.argv[1] if len(sys.argv) > 1 else DEFAULT_MINT
    print(f"\n=== Solana Devnet RPC Token Query ===")
    print(f"Mint: {mint}\n")

    # 1. Supply
    print("1. Supply...")
    info = rpc("getAccountInfo", [mint, {"encoding": "jsonParsed"}])
    p = info["value"]["data"]["parsed"]["info"]
    supply = int(p["supply"]) / 10 ** p["decimals"]
    print(f"   Decimals:   {p['decimals']}")
    print(f"   Supply:     {supply:,.2f}")

    # 2. Largest holders
    print("\n2. Largest holders...")
    time.sleep(0.5)
    try:
        largest = rpc("getTokenLargestAccounts", [mint])
        for i, a in enumerate(largest.get("value", [])[:5]):
            bal = float(a.get("uiAmount", 0))
            print(f"   [{i+1}] {a['address'][:12]}...  {bal:,.2f}")
    except Exception as e:
        print(f"   Skipped: {e}")

    # 3. Recent signatures
    print("\n3. Recent transaction signatures...")
    time.sleep(0.5)
    try:
        sigs = rpc("getSignaturesForAddress", [mint, {"limit": 5}])
        print(f"   Found: {len(sigs)} sigs")
        for s in sigs[:3]:
            print(f"   {s['signature'][:20]}...  slot={s['slot']}")
    except Exception as e:
        print(f"   {e}")

    print("\n=== Done ===")


if __name__ == "__main__":
    main()
